content/articles/2019-02-purrr-0-3-0.Rmd

```diff
@@ -0,0 +1,308 @@
+
+```{r, eval = FALSE}
+# Install the latest version with:
+install.packages("purrr")
+
+# Start working with purrr:
+library("purrr")
+```
+
+In this release, `pluck()` gets a few new variants:
+
+* `pluck<-`, `assign_in()` and `modify_in()` allow deep modifications of nested structures.
+
+* `chuck()` is a stricter variant of `pluck()` that consistently fails instead of returning `NULL` when the structure does not have the expected shape.
+
+A new set of tools helps limiting the rate at which a function is called:
+
+* `slowly()` forces a function to sleep between each invokation.
+
+* `insistently()` automatically invokes a function again on error until it succeeds, and sleeps between invokations.
+
+* The `rate_delay()` and `rate_backoff()` helpers control the invokation rate of `slowly()` and `insistently()`.
+
+Finally, the consistency of the interface has been polished:
+
+* The direction of iteration/application is now consistently specified with a `.dir` argument.
+
+* Many missing functions were added to fill the gaps: `accumulate2()`, `imodify()`, `map_depth()`, ...
+
+* `partial()` has a much improved and more flexible interface.
+
+Find a detailed account of the changes in the [NEWS](https://github.com/tidyverse/purrr/blob/master/NEWS.md#purrr-030) file.
+
+
+## New pluck variants
+
+`pluck()` implements a generalised form of `[[` that allow you to index deeply and flexibly into data structures. For instance, `pluck(x, "foo", 2)` is equivalent to `x[["foo"]][[2]]`. purrr 0.3.0 introduces variants of `pluck()` to make it easier to work with deep data structures.
+
+
+### Pluck assignment
+
+This release introduces the new functions `pluck<-`, `assign_in()` and `modify_in()` as assignment variants of `pluck()`.
+
+To illustrate deep assignment, Let's create a nested data structure:
+
+```{r}
+x <- list(foo = list(1, 2), bar = list(3, 4))
+str(x)
+```
+
+This sort of repeated structure is the kind of data where `pluck()` shines:
+
+```{r}
+pluck(x, "foo", 2)
+
+pluck(x, "bar", 1)
+```
+
+You can now use the same syntax to modify the data:
+
+```{r}
+pluck(x, "foo", 2) <- 100
+str(x)
+```
+
+`pluck<-` also has a functional form that does not modify objects in your environment, but instead returns a modified copy:
+
+```{r}
+out <- assign_in(x, list("foo", 2), 2000)
+
+# The object is still the same as before
+str(x)
+
+# The modified data is in `out`
+str(out)
+```
+
+Finally, `modify_in()` is a variant of `modify()` that only changes the pluck location with the result of applying a function:
+
+```{r}
+out <- modify_in(x, list("foo", 2), as.character)
+str(out)
+```
+
+
+### Stricter pluck()
+
+Thanks to Daniel Barnett (@daniel-barnett on Github), `pluck()` now has a stricter cousin `chuck()`. Whereas `pluck()` is very permissive regarding non-existing locations and returns `NULL` in these cases, `chuck()` fails consistently with informative messages (i.e., it "chucks" an error message):
```

Comment:
Whereas `pluck()` ..., and `[[` inconsistently returns `NULL`, `NA` or throws an error, `chuck()` ...